x/configurl/shadowsocks.go

```diff
@@ -25,52 +26,61 @@ import (
-	dialer, err := shadowsocks.NewStreamDialer(endpoint, config.cryptoKey)
-	if err != nil {
-		return nil, err
-	}
-	if len(config.prefix) > 0 {
-		dialer.SaltGenerator = shadowsocks.NewPrefixSaltGenerator(config.prefix)
-	}
-	return dialer, nil
+func registerShadowsocksStreamDialer(r TypeRegistry[transport.StreamDialer], typeID string, newSD BuildFunc[transport.StreamDialer]) {
+	r.RegisterType(typeID, func(ctx context.Context, config *Config) (transport.StreamDialer, error) {
+		sd, err := newSD(ctx, config.BaseConfig)
+		if err != nil {
+			return nil, err
+		}
+		ssConfig, err := parseShadowsocksURL(config.URL)
+		if err != nil {
+			return nil, err
+		}
+		endpoint := &transport.StreamDialerEndpoint{Dialer: sd, Address: ssConfig.serverAddress}
+		dialer, err := shadowsocks.NewStreamDialer(endpoint, ssConfig.cryptoKey)
+		if err != nil {
+			return nil, err
+		}
+		if len(ssConfig.prefix) > 0 {
+			dialer.SaltGenerator = shadowsocks.NewPrefixSaltGenerator(ssConfig.prefix)
+		}
+		return dialer, nil
+	})
 }
 
-func wrapPacketDialerWithShadowsocks(_ func() (transport.StreamDialer, error), innerPD func() (transport.PacketDialer, error), configURL *url.URL) (transport.PacketDialer, error) {
-	pd, err := innerPD()
-	if err != nil {
-		return nil, err
-	}
-	config, err := parseShadowsocksURL(configURL)
-	if err != nil {
-		return nil, err
-	}
-	endpoint := &transport.PacketDialerEndpoint{Dialer: pd, Address: config.serverAddress}
-	listener, err := shadowsocks.NewPacketListener(endpoint, config.cryptoKey)
-	if err != nil {
-		return nil, err
-	}
-	dialer := transport.PacketListenerDialer{Listener: listener}
-	return dialer, nil
+func registerShadowsocksPacketDialer(r TypeRegistry[transport.PacketDialer], typeID string, newPD BuildFunc[transport.PacketDialer]) {
+	r.RegisterType(typeID, func(ctx context.Context, config *Config) (transport.PacketDialer, error) {
+		pd, err := newPD(ctx, config.BaseConfig)
+		if err != nil {
+			return nil, err
+		}
+		ssConfig, err := parseShadowsocksURL(config.URL)
+		if err != nil {
+			return nil, err
+		}
+		endpoint := &transport.PacketDialerEndpoint{Dialer: pd, Address: ssConfig.serverAddress}
+		pl, err := shadowsocks.NewPacketListener(endpoint, ssConfig.cryptoKey)
+		if err != nil {
+			return nil, err
+		}
+		// TODO: support UDP prefix.
+		return transport.PacketListenerDialer{Listener: pl}, nil
+	})
 }
 
-func newShadowsocksPacketListenerFromURL(configURL *url.URL) (transport.PacketListener, error) {
-	config, err := parseShadowsocksURL(configURL)
-	if err != nil {
-		return nil, err
-	}
-	// TODO: accept an inner dialer from the caller and pass it to UDPEndpoint
-	ep := &transport.UDPEndpoint{Address: config.serverAddress}
-	return shadowsocks.NewPacketListener(ep, config.cryptoKey)
+func registerShadowsocksPacketListener(r TypeRegistry[transport.PacketListener], typeID string, newPD BuildFunc[transport.PacketDialer]) {
```

Comment:
Not for this PR, but I think we can replace all `PacketListener` with the `PacketDialer` for Shadowsocks in the future. Are we doing anything special in the `PacketListener`?